#### **Библиотека BeautifulSoup**

Для поиска необходимых нам данных мы будем использовать библиотеку **BeautifulSoup,** которая позволяет по названию тегов и их атрибутов получать содержащийся в них текст.

**BeautifulSoup** не является частью стандартной библиотеки, поэтому для начала её нужно установить.

После установки импортируем библиотеку в наш код:

In [1]:
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

Теперь мы можем извлекать данные из любой веб-страницы.

Ранее мы уже получили содержимое страницы с помощью *GET-запроса* и сохранили информацию в переменной **response,** теперь создадим объект **BeautifulSoup** с именем **page,** указывая в качестве параметра **html.parser.**

Для примера получим информацию o **title** (с англ. **заголовок**) - это строка, которая отображается на вкладке браузера:

In [2]:
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://nplus1.ru/news/2021/10/11/econobel2021' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
print(page.title) # Получаем тег title, отображающийся на вкладке браузера
print(page.title.text) # Выводим текст из полученного тега, который содержится в атрибуте text

<title>Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей</title>
Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей


Если при запросе к сайту, а затем при его разборе с помощью **BeautifulSoup** в тексте страницы не находится нужный тег, попробуйте вывести на печать пару тысяч символов текста страницы. Если там обнаружится нечто похожее на капчу, возможно, сайт посчитал вас роботом и отказывается выдавать содержимое. Чтобы получить его, попробуйте «притвориться» браузером при запросе из скрипта:

In [3]:
requests.get(url, headers={'User-Agent': 'Mozilla/5.0'})

<Response [200]>

##### **<center> ИЗВЛЕКАЕМ ЗАГОЛОВОК И ВРЕМЯ НАПИСАНИЯ СТАТЬИ </center>**

Выполним поставленную ранее задачу: получить информацию о странице и извлечь заголовок статьи, опубликованной на этой странице, дату публикации, а также текст статьи.

Предположим, что мы знаем, что в *HTML-коде* рассматриваемой нами страницы заголовок статьи заключён в тег **<*h1*>** … **<*/h1*>** (заголовок первого уровня).

Тогда мы можем получить его текст с помощью метода **find()** (с англ. **найти**) объекта **BeautifulSoup,** передав ему название интересующего нас тега:

In [4]:
# Применяем метод find() к объекту и выводим результат на экран
print(page.find('h1').text) 


            Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей
          


Но как же узнать, в каких именно тегах заключена необходимая информация?

Проще всего это сделать с помощью так называемого инструмента разработчика, который есть во всех современных браузерах. Покажем, как открыть данный инструмент на примере использования браузера *Google Chrome.*

Устанавливаем курсор на элементе страницы (заголовок статьи), информацию о котором хотим получить, нажимаем на правую клавишу мыши и в выпадающем списке выбираем пункт **«Просмотреть код элемента»** или **«Посмотреть код»** в зависимости от браузера.

##### **Задача 1**

Напишите функцию **wiki_header,** которая по адресу страницы возвращает заголовок первого уровня для статей на *Wikipedia.*

Функция **wiki_header** принимает один аргумент - **url.**

In [9]:
wiki_header('https://en.wikipedia.org/wiki/Operating_system')
'Operating system'

import requests
from bs4 import BeautifulSoup

def wiki_header(url):
    page=BeautifulSoup(requests.get(url).text,'html.parser')
    header=page.find('h1').text
    return header

##### **<center> НЕУНИКАЛЬНЫЕ ТЕГИ: ИЗВЛЕКАЕМ ТЕКСТ И ДАТУ ПУБЛИКАЦИИ СТАТЬИ </center>**

Теперь получим сам текст статьи. Как вы уже знаете, первым делом необходимо определить, в какой тег он заключён. Применим, как и ранее, инструмент разработчика.

Видим, что искомый текст заключён в тег **<*div*>** … **<*/div*>**. Попробуем извлечь его уже известным нам способом - с помощью метода **find()** - и выведем его на экран.

In [10]:
print(page.find('div').text) # Выводим содержимое атрибута text тега div

Мы увидели не то, что ожидали - кучу текста, не имеющего отношения к тому, что мы искали...

*В чём же проблема?*

Дело в том, что теги **<*div*>** … **<*/div*>** очень распространённые и на странице их очень много. Метод **find()** нашёл первый из них, но это не то, что нам надо.

Посмотрим на нашу страницу, используя инструмент разработчика, ещё раз. Можем заметить, что у искомого текста есть свой класс - **n1_material text-18:**

Передадим название класса в метод **find()** с помощью аргумента **class_** и получим текст статьи:

In [14]:
print(page.find('div', class_='n1_material text-18').text) # Выводим содержимое атрибута text тега div класса n1_material text-18

Премия Шведского национального банка по экономическим наукам памяти Альфреда Нобеля за 2021 год присуждена Дэвиду Карду (David Card) за его вклад в эмпирические исследования экономики рынка труда, а также Джошуа Энгристу (Joshua Angrist) и Гвидо Имбенсу (Guido Imbens) за их вклад в методологию анализа причинно-следственных связей. Прямая трансляция церемонии объявления лауреатов шла на официальном сайте Нобелевской премии.


В данном случае происходит поиск точного строкового значения **class** атрибута, т. е. выполнение строк кода:

In [16]:
print(page.find('div', class_='n1_material').text)

Премия Шведского национального банка по экономическим наукам памяти Альфреда Нобеля за 2021 год присуждена Дэвиду Карду (David Card) за его вклад в эмпирические исследования экономики рынка труда, а также Джошуа Энгристу (Joshua Angrist) и Гвидо Имбенсу (Guido Imbens) за их вклад в методологию анализа причинно-следственных связей. Прямая трансляция церемонии объявления лауреатов шла на официальном сайте Нобелевской премии.


In [17]:
print(page.find('div', class_='n1_material text-18').text)

Премия Шведского национального банка по экономическим наукам памяти Альфреда Нобеля за 2021 год присуждена Дэвиду Карду (David Card) за его вклад в эмпирические исследования экономики рынка труда, а также Джошуа Энгристу (Joshua Angrist) и Гвидо Имбенсу (Guido Imbens) за их вклад в методологию анализа причинно-следственных связей. Прямая трансляция церемонии объявления лауреатов шла на официальном сайте Нобелевской премии.


даст одинаковый результат.

При выполнении строки кода

In [18]:
print(page.find('div', class_='text-18 n1_material').text)

AttributeError: 'NoneType' object has no attribute 'text'

мы получим ошибку, так как такого строкового значения в области поиска нет.

Аналогично получим информации о теге, который содержит дату написания статьи, отображаемую в левом верхнем углу страницы.

Итак, нам нужен тег **<*a*>** … **<*/a*>** с классом **"relative before:block before:w-px before:bg-current before:h-4 before:absolute before:left-0 group pl-2 flex inline-flex items-center".** Для поиска достаточно указать в качестве класса **"relative",** отбросив дополнительные настройки.

Теперь получим данные из него с помощью уже известного метода **find(),** передав название нужного тега:

In [19]:
# Выводим на экран содержимое атрибута text тега a с классом "relative"
print(page.find('a', class_= "relative").text)


11.10.21



Задача решена - мы извлекли из контента страницы заголовок статьи, опубликованной на странице, дату публикации, а также текст статьи.

##### **<center> СБОР НЕСКОЛЬКИХ ЭЛЕМЕНТОВ: СОБИРАЕМ ВСЕ ССЫЛКИ НА СТРАНИЦЕ </center>**

Рассмотрим ещё один сценарий: вы хотите собрать сразу несколько элементов со страницы. Например, представьте, что вы хотите получить названия всех языков программирования, упомянутых на странице в *Wikipedia* в статье про языки программирования.

Можно заметить, что все названия языков программирования на этой странице связаны ссылками c соответствующими статьями о них. Таким образом, нам необходимо собрать все ссылки на странице. Для ссылок в *HTML* предусмотрен тег **<*a*>** … **<*/a*>**. Попробуем использовать **find():**

In [20]:
url = 'https://en.wikipedia.org/wiki/List_of_programming_languages' # Задаём адрес ресурса
response = requests.get(url) # Делаем GET-запрос к ресурсу
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup
print(page.find('a')) # Ищем ссылку по тегу <a> и выводим её на экран

<a class="mw-jump-link" href="#bodyContent">Jump to content</a>


Мы получили только одну ссылку, хотя на странице их явно больше.

Это происходит, потому что метод **find()** возвращает только первый подходящий элемент. Если требуется получить больше элементов, необходимо воспользоваться методом **find_all()** (с англ. **найти все**)**:**

In [21]:
links = page.find_all('a') # Ищем все ссылки на странице и сохраняем в переменной links в виде списка
print(len(links)) # Выводим количество найденных ссылок

958


Итак, на момент создания этих учебных материалов на странице содержалось **958** ссылок. Посмотрим на некоторые из них:

In [22]:
print([link.text for link in links[500:510]]) # Выводим ссылки с 500 по 509 включительно

['Machine code', 'MAD', 'MAD/I', 'Magik', 'Magma', 'Maple', 'MAPPER', 'MARK-IV', 'Mary', 'MATLAB']


Не все ссылки соответствуют названиям языков программирования - страница содержит также «служебные» ссылки, такие, например, как **Jump to navigation** (с англ. *Перейти к навигации*) или **Alphabetical** (с англ. *По алфавиту*):

In [23]:
print([link.text for link in links[0:10]]) # Выводим ссылки с 1 по 9 включительно

['Jump to content', 'Main page', 'Contents', 'Current events', 'Random article', 'About Wikipedia', 'Contact us', 'Donate', 'Help', 'Learn to edit']


Для обработки полученных данных и исключения «лишней» информации можно, например, использовать подходы, которые вы изучили в модуле *PYTHON-14.*

В заключение заметим, что **BeautifulSoup** - достаточно мощная библиотека. Мы рассмотрели её базовые возможности, но их полный список гораздо шире.